hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -2467,27 +2468,66 @@ def test_not_equal(x, y):
-        """Attempts to replace most bytes in the buffer with 0 or 1. The main
+        """Attempts to minimize the "alphabet" - the set of bytes that
+        are used in the representation of the current buffer. The main
         benefit of this is that it significantly increases our cache hit rate
         by making things that are equivalent more likely to have the same
-        representation.
-
-        We only run this once rather than as part of the main passes as
-        once it's done its magic it's unlikely to ever be useful again.
-        It's important that it runs first though, because it makes
-        everything that comes after it faster because of the cache hits.
+        representation, but it's also generally a rather effective "fuzzing"
+        step that gives us a lot of good opportunities to slip to a smaller
+        representation of the same bug.
         """
-        for c in (1, 0):
-            alphabet = set(self.buffer) - set(hrange(c + 1))
+        for c in range(255, 0, -1):
+            buf = self.buffer
 
-            if not alphabet:
+            if c not in buf:
                 continue
 
-            def clear_to(reduced):
-                reduced = set(reduced)
-                attempt = hbytes(
-                    [b if b <= c or b in reduced else c for b in self.buffer]
+            def try_replace(d):
+                if d < 0:
+                    return False
+
+                return self.consider_new_buffer(
+                    bytes([d if b == c else b for b in buf])
                 )
-                return self.consider_new_buffer(attempt)
 
-            Length.shrink(sorted(alphabet), clear_to, random=self.random)
+            # If we cannot replace the current byte with its predecessor,
+            # assume it is already minimal and continue on. This ensures
+            # we make no more than one call per distinct byte value in the
+            # event that no shrinks are possible here.
+            if not try_replace(c - 1):
+                continue
+
+            # We now try replacing with 0 and 1. If this works we attempt
+            # a bulk replacement where we replace all values which are close
+            # to c but smaller with 0 or 1 as well. This helps us substantially
+            # in cases where we have a lot of "dead" bytes that don't really do
+            # much, as it allows us to replace many of them in one go rather
+            # than one at a time. An example of where this matters is
+            # test_minimize_multiple_elements_in_silly_large_int_range_min_is_not_dupe
+            # in test_shrink_quality.py
+            def bulk_replace(d):
+                if try_replace(d):
+
+                    def replace_range(k):
+                        if k > c:
+                            return False
+                        return self.consider_new_buffer(
+                            bytes([d if c - k <= b <= c and b > d else b for b in buf])
```

Comment:
Pulling this condition out into a named function could help a lot. Something like:

```python
def should_replace_byte(b):
    return c - k <= b <= c and d < b
return self.consider_new_buffer(
    hbytes([d if should_replace_byte(b) else b for b in buf])
)
```

The single-letter variables are still making my head spin a bit, but at least they're not blending into the replacement comprehension.